Make TTreeReaderArray's iterator random access

bluehood opened the pull request.

bluehood commented:
The `colWiseRead` tests are fixed by [this PR in roottest](https://github.com/root-project/roottest/pull/102).

bluehood commented:
Rebased on master and fixed almost all of the clang-format issues.

bluehood commented:
On a second thought this still needs some tweaks: ideally when `TTreeReaderArray` is const its iterator will behave like a const iterator, otherwise not.

dpiparo commented:
@phsft-bot build please

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
+   const T &operator[](std::size_t idx) const { return At(idx); }
+
+   /// Random access iterator to the elements of a TTreeReaderArray
+   class Iterator_t : public std::iterator<std::random_access_iterator_tag, T> {
```
I prefer to keep the order: types, data members, function members. That's because fairly often the types are needed by the members, and having a common order helps readability (the long version of "our (future) coding rules say so).

bluehood replied:
changed the order of TTreeReaderArray internals to Iterator_t definition, using declarations, methods.

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
+   const T &operator[](std::size_t idx) const { return At(idx); }
+
+   /// Random access iterator to the elements of a TTreeReaderArray
+   class Iterator_t : public std::iterator<std::random_access_iterator_tag, T> {
```
`std::iterator` is actually deprecated. I'm not requesting a change - I'm suggesting it in case you think it helps this PR. See e.g. https://stackoverflow.com/questions/43268146/why-is-stditerator-deprecated

bluehood replied:
I saw it was deprecated in c++17 and did not think much of it (I thought something else changed in c++17 that allowed its deprecation). After reading the stackoverflow thread you linked, I removed `std::iterator` and just declared the required nested types myself.

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
-      return IsEmpty() ? Iterator_t() : Iterator_t(0, this);
-   }
-   Iterator_t end() const { return Iterator_t(); }
+   iterator begin() { return IsEmpty() ? Iterator_t() : Iterator_t(0u, this); }
```
`Iterator_t(0u, this)` is already checking for `0 >= GetSize()`; can you remove the `IsEmpty()` check? (Sorry, I do realize the original code had this, too?)

bluehood replied:
Done

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
-   Iterator_t end() const { return Iterator_t(); }
+   iterator begin() { return IsEmpty() ? Iterator_t() : Iterator_t(0u, this); }
+   iterator end() { return Iterator_t(GetSize(), this); }
+   const_iterator begin() const { return IsEmpty() ? Iterator_t() : Iterator_t(0u, this); }
```
Can this please return `cbegin()`?

bluehood replied:
done

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
+   iterator begin() { return IsEmpty() ? Iterator_t() : Iterator_t(0u, this); }
+   iterator end() { return Iterator_t(GetSize(), this); }
+   const_iterator begin() const { return IsEmpty() ? Iterator_t() : Iterator_t(0u, this); }
+   const_iterator end() const { return Iterator_t(GetSize(), this); }
```
Can this please return `cend()`?

bluehood replied:
done

Axel-Naumann commented on tree/treeplayer/test/array.cxx:
```diff
@@ -66,3 +67,138 @@ TEST(TTreeReaderArray, MultiReaders) {
    }
 }
+
+TEST(TTreeReaderArray, Iterator) {
```
Sweeet! Thanks! Should this be moved into iterval.cxx? I'd say it has acquired critical mass...

bluehood replied:
As you prefer, but I'd opt for a more verbose name...would `readerarray_iterator.cxx` be ok?

Axel-Naumann commented on tree/treeplayer/test/array.cxx:
```diff
@@ -66,3 +67,138 @@ TEST(TTreeReaderArray, MultiReaders) {
+   EXPECT_GE(b + 1, b + 1);
+
+   // difference
+   const auto length = arr.end() - arr.begin();
```
Can you use the actual type here, as that is part of the iterator pattern? Just to make sure that if people rely on it being `std::ptrdiff_t` it actually compiles without a warning.

bluehood replied:
Done

Axel-Naumann commented on tree/treeplayer/test/array.cxx:
```diff
@@ -66,3 +67,138 @@ TEST(TTreeReaderArray, MultiReaders) {
+   // ranged for loop
+   int truth = 0;
+   for (auto v : arr)
+      EXPECT_DOUBLE_EQ(v, truth++);
```
Why `DOUBLE`?

bluehood replied:
Missed refactoring, fixed

Axel-Naumann commented on tree/treeplayer/test/array.cxx:
```diff
@@ -66,3 +67,138 @@ TEST(TTreeReaderArray, MultiReaders) {
+}
+
+TEST(TTreeReaderArray, ConstIterator) {
+   // test TTreeReaderArray's nested iterator type when TTreeReaderArray is const
```
Is there any chance that you can combine these two tests? There seems to be a lot of code duplication, and I suppose the `const_iterator` functionality should be a subset of the `iterator` functionality.

I find gtest works quite well even if the test calls functions that fail.

bluehood replied:
Once `const_iterator` is fixed, the matrix will be `(iterator, const_iterator)x(TTreeReaderArray, const TTreeReaderArray)`. How would you like to divide the cases in unit tests? I would actually write 4 unit tests, but I suspect you'd prefer 2..?

Given that I can think of ways to reduce code duplication

Axel-Naumann replied:
Well I think the two combinations `(TTreeReaderArray, iterator)` and `(const TTreeReaderArray, const_iterator)` are just fine - but maybe that's just me not being eager enough ;-)

bluehood replied:
now testing for this two cases

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
+      {
+         auto ret = *this;
+         auto &self = *this;
+         ++self;
```
You find these two lines nicer than `this->operator++()` or if need be `++(*this)`? I find the option you chose my third favorite, due to the `auto &` and the second local var that seems to be almost identical to `ret`. I.e. I find this code much more complex to grasp due to the extra degree of freedom (extra local var)... But as "I find" tries to say: this is really a matter of taste, so if you disagree then feel free to keep yours!

bluehood replied:
I agree, changed

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
+   class Iterator_t : public std::iterator<std::random_access_iterator_tag, T> {
+      TTreeReaderArray *fArray; ///< The array iterated over; nullptr if invalid/past-the-end.
+      std::size_t fIndex;       ///< Current index in the array.
+      std::size_t fSize;        ///< Size of the TTreeReaderArray
```
Do we know that the iterator is invalidated when a new entry is loaded? I guess that's fine, right? ('Cause before, the size was implicitly allowed to change.)

bluehood replied:
uhm, what do you expect to happen if I get an iterator, possibly use it, then load a new entry and try to keep using it?
should we support this usage? should we not, and warn users that do it?

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
+      }
 
-   typedef Iterator_t iterator;
+      bool operator<(const Iterator_t &other) const { return fIndex < other.fIndex; }
```
Does `end() < end()` behave as expected?

bluehood replied:
it should return `false` since both `fIndex` and `other.fIndex` are equal to `fSize`. Unless you meant something more subtle..?

Axel-Naumann commented on tree/treeplayer/inc/TTreeReaderArray.h:
```diff
@@ -59,90 +59,159 @@ namespace Internal {
-   T& At(size_t idx) { return *(T*)UntypedAt(idx); }
-   T& operator[](size_t idx) { return At(idx); }
+   using iterator = Iterator_t;
+   using const_iterator = const Iterator_t;
```
Doesn't that mean that this is ill-formed:
```
auto i = arr.cbegin();
i += 3;
```
The "const" part relates to modifying the elements, not the iterator.

bluehood replied:
As per my comment below, I got `const_iterator` wrong. In the process of fixing it :)

bluehood commented:
Hi @axel-naumann
thank you for the thorough review!
I will address all comments when I'm back on Wednesday.

A major problem I see is that I have to fix the behaviour of the const iterator. Are const iterators usually implemented as a different type altogether? Should I add a boolean template parameter that causes changes in the return type of `operator*`?

Axel-Naumann commented:
The trick is basically to template on the pointer type, so you can pass it as `T` for `iterator` and `const T` for `const_iterator`.

bluehood commented:
Cleaned up commit history, rebased and pushed again.
All previous comments should have been addressed except avoiding code duplication in the unit test.

I have also introduced `iterator operator+(int, iterator)`, which cppreference lists as required.
Also, now incrementing an invalid (e.g. past-the-end) iterator is a no-op.

bluehood commented:
@Axel-Naumann I will merge after refactoring the tests then.

One topic that I am not addressing in the PR is what should happen if the user gets an iterator, possibly uses it, then loads a new entry and tries to keep using the old iterator.
We could document that iterators on a certain entry are to be considered invalidated when the entry is switched, but _actually_ invalidating the iterators would be a bit cumbersome (and beyond the scope of this PR).

bluehood commented:
Tests has been augmented and refactored to avoid code duplication.

There's one issue that I'm not sure how to solve: random access iterators are required to support the forms `it + n` as well as `n + it` (where `n` in our case is of type `std::ptrdiff_t`. The second form requires the definition of a free-standing `operator+`, which I implemented this way:

```c++
template <typename T>
auto operator+(std::ptrdiff_t n, const typename TTreeReaderArray<T>::iterator &it)
-> decltype(it + n)
{
   return it + n;
}
```

Problem: in the definition above `T` cannot be deduced from the dependent type `iterator`, so this overload is never considered. Any idea how to work around this?
